### PR TITLE
Fix bug in differentiable Tpfa with mpfa as base discretization

### LINKED ISSUE #1556: Differentiable TPFA with MPFA as base discretization
Said combination produces weird results at the boundaries, at least if gravity is present. See below for an illustration. ATM, I have done no further investigation and, unfortunately, not produced anything close to an MRE (believe me, the simulation I ran is not minimal).

Image

### PR DESCRIPTION
**Pull request by keileg**
## Proposed changes
This PR fixes two minor bugs related to the differentiable two-point flux approximation. In both cases, terms related to the discretization of boundary conditions were missing, with the scheme using mpfa as base discretization was mainly suffering.

The bugfix revealed a mistaken assumption in a test related to the sparsity of Jacobian matrices on faces with Neumann boundary conditions. The tested condition is only true when using tpfa as a base discretization, though the bug had made the test pass also for mpfa. The test now only runs for tpfa.

Also added is a test that verifies that the number of Newton iterations with the differentiable scheme is not larger than the number with standard discretizations. In practice, the number of iterations is the same for most if not all of the test parametrizations. I made some attempts at finding a permeability function that gave a strict improvement with the differentiable scheme, but did not succeed. This does not imply that such a function does not exist though, so I am open for a discussion on whether to try once more. _Please consider this question in particular in the review._

Resolves #1556.

## Types of changes

What types of changes does this PR introduce to PorePy?
_Put an `x` in the boxes that apply._

- [ ] Minor change (e.g., dependency bumps, broken links).
- [ ] Bugfix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] Testing (contribution related to testing of existing or new functionality).
- [ ] Documentation (contribution related to adding, improving, or fixing documentation).
- [ ] Maintenance (e.g., improve logic and performance, remove obsolete code).
- [ ] Other:

## Checklist

_Put an `x` in the boxes that apply or explain briefly why the box is not relevant._

- [ ] The documentation is up-to-date.
- [ ] Static typing is included in the update.
- [ ] This PR does not duplicate existing functionality.
- [ ] The update is covered by the test suite (including tests added in the PR).
- [ ] If new skipped tests have been introduced in this PR, `pytest` was run with the `--run-skipped` flag.

### CODE REVIEW

File: tests/numerics/fv/test_tpfa.py [hunk 1]
```diff
@@ -768,6 +772,146 @@ def test_diff_tpfa_and_standard_tpfa_give_same_linear_system(base_discr: str):
...
+@pytest.mark.parametrize("constant_permeability", [True, False])
+def test_diff_tpfa_newton_performance(
+    base_scheme: Literal["mpfa", "tpfa"],
+    gravity: bool,
+    include_fracture: bool,
+    constant_permeability: bool,
+):
+    """Verify that the solutions computed with and without diff-tpfa are consistent,
+    and that inclusion of the diffenetiable tpfa does not increase the number of
+    Newton iteration.
```
Comment (IvarStefansson) [inline tests/numerics/fv/test_tpfa.py]:
```suggestion
    Newton iterations.
```

**Review by IvarStefansson — APPROVED** [review]
I don't think it's worth it atm to spend more time finding a "better" case to test.
